File: alphago_tictactoe.py
"""
AlphaGo 简化版 —— 井字棋
========================
核心组件：
1. 棋盘逻辑（Board）
2. 神经网络（PolicyValueNet）—— 策略 + 价值
3. 蒙特卡洛树搜索（MCTS）—— 用神经网络指导搜索
4. 自我对弈训练（train）—— 通过实战提升棋力

运行: pip install torch numpy && python alphago_tictactoe.py
"""

import math
import random
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from collections import defaultdict

# ============================================================
# 第一部分：棋盘逻辑
# ============================================================

class Board:
    """井字棋棋盘"""

    def __init__(self):
        # 3x3 棋盘，0=空，1=黑棋(先手)，-1=白棋(后手)
        self.grid = np.zeros(9, dtype=np.float32)
        self.current_player = 1  # 黑棋先走

    def copy(self):
        b = Board()
        b.grid = self.grid.copy()
        b.current_player = self.current_player
        return b

    def legal_moves(self):
        """返回所有合法落子位置"""
        return [i for i in range(9) if self.grid[i] == 0]

    def play(self, pos):
        """在 pos 位置落子"""
        assert self.grid[pos] == 0, f"位置 {pos} 已有棋子"
        self.grid[pos] = self.current_player
        self.current_player *= -1  # 换手

    def check_winner(self):
        """检查胜负：返回 1(黑赢), -1(白赢), 0(平局), None(未结束)"""
        lines = [
            [0,1,2], [3,4,5], [6,7,8],  # 横
            [0,3,6], [1,4,7], [2,5,8],  # 竖
            [0,4,8], [2,4,6],            # 对角
        ]
        for line in lines:
            s = self.grid[line[0]] + self.grid[line[1]] + self.grid[line[2]]
            if s == 3: return 1    # 黑赢
            if s == -3: return -1  # 白赢
        if len(self.legal_moves()) == 0:
            return 0  # 平局
        return None  # 未结束

    def to_tensor(self):
        """
        把棋盘转成神经网络的输入
        3个通道：
          通道0: 我的棋子位置（1=有子）
          通道1: 对手棋子位置（1=有子）
          通道2: 当前玩家标识（全1或全-1）
        """
        my_pieces = (self.grid == self.current_player).astype(np.float32)
        opp_pieces = (self.grid == -self.current_player).astype(np.float32)
        player_plane = np.full(9, self.current_player, dtype=np.float32)
        return np.concatenate([my_pieces, opp_pieces, player_plane])

    def display(self):
        """打印棋盘"""
        symbols = {0: '.', 1: 'X', -1: 'O'}
        for row in range(3):
            print(' '.join(symbols[self.grid[row*3 + col]] for col in range(3)))
        print()


# ============================================================
# 第二部分：神经网络 —— 策略头 + 价值头
# ============================================================

class PolicyValueNet(nn.Module):
    """
    输入：棋盘状态 (3 x 9)
    输出：
      - policy: 每个位置的落子概率 (9个数字)
      - value:  当前局面的胜率评估 (1个数字, -1到1)
    """

    def __init__(self):
        super().__init__()
        # 共享的特征提取层
        self.shared = nn.Sequential(
            nn.Linear(27, 128),   # 3通道 x 9格 = 27维输入
            nn.ReLU(),
            nn.Linear(128, 128),
            nn.ReLU(),
        )
        # 策略头：输出每个位置的落子概率
        self.policy_head = nn.Linear(128, 9)
        # 价值头：输出当前局面胜率
        self.value_head = nn.Linear(128, 1)

    def forward(self, x):
        features = self.shared(x)
        policy_logits = self.policy_head(features)  # 未归一化的概率
        value = torch.tanh(self.value_head(features))  # 压缩到 [-1, 1]
        return policy_logits, value


# ============================================================
# 第三部分：蒙特卡洛树搜索（MCTS）
# ============================================================

class MCTSNode:
    """搜索树的一个节点"""

    def __init__(self, parent=None, move=None):
        self.parent = parent
        self.move = move           # 到达这个节点的走法
        self.children = {}         # move -> MCTSNode
        self.visits = 0            # 访问次数
        self.value_sum = 0.0       # 累计价值
        self.prior = 0.0           # 神经网络给的先验概率

    @property
    def value(self):
        """平均价值"""
        return self.value_sum / self.visits if self.visits > 0 else 0

    def ucb_score(self, c_puct=1.4):
        """
        UCB 公式：平衡「探索」和「利用」
        - value: 这个节点的平均胜率（利用）
        - prior * sqrt(parent_visits) / (1 + visits): 先验概率引导的探索
        """
        if self.visits == 0:
            exploration = float('inf')
        else:
            parent_visits = self.parent.visits if self.parent else 1
            exploration = c_puct * self.prior * math.sqrt(parent_visits) / (1 + self.visits)
        return self.value + exploration


def mcts_search(board, net, num_simulations=100):
    """
    对一个局面做 MCTS 搜索

    核心循环：
    1. 选择（Selection）：从根节点沿 UCB 最高的路径走到叶子
    2. 扩展（Expansion）：用神经网络评估叶子节点
    3. 回溯（Backpropagation）：把评估值传回路径上所有节点

    返回：每个合法走法的访问次数分布（作为改进后的策略）
    """
    root = MCTSNode()

    for _ in range(num_simulations):
        node = root
        sim_board = board.copy()

        # --- 1. 选择：沿 UCB 最高的路径走到叶子 ---
        while node.children and sim_board.check_winner() is None:
            # 选 UCB 最高的子节点
            best_move = max(node.children, key=lambda m: node.children[m].ucb_score())
            node = node.children[best_move]
            sim_board.play(best_move)

        # --- 2. 扩展：评估叶子节点 ---
        winner = sim_board.check_winner()
        if winner is not None:
            # 游戏已结束，直接用真实结果
            # winner == 当前玩家 → +1，对手赢 → -1，平局 → 0
            leaf_value = winner * sim_board.current_player * (-1)
            # 注意：这里 winner 是相对于原始玩家的，需要转换
            if winner == 0:
                leaf_value = 0
            elif winner == board.current_player:
                leaf_value = 1
            else:
                leaf_value = -1
        else:
            # 用神经网络评估
            state = torch.FloatTensor(sim_board.to_tensor()).unsqueeze(0)
            with torch.no_grad():
                policy_logits, value = net(state)

            leaf_value = value.item()

            # 创建子节点，用神经网络的策略作为先验概率
            policy = torch.softmax(policy_logits, dim=1).squeeze().numpy()
            legal = sim_board.legal_moves()

            # 归一化合法走法的概率
            legal_policy = {m: policy[m] for m in legal}
            total = sum(legal_policy.values())
            if total > 0:
                legal_policy = {m: p/total for m, p in legal_policy.items()}
            else:
                legal_policy = {m: 1/len(legal) for m in legal}

            for move in legal:
                child = MCTSNode(parent=node, move=move)
                child.prior = legal_policy[move]
                node.children[move] = child

        # --- 3. 回溯：把叶子价值传回根节点 ---
        # 注意：每上一层要翻转符号（对手视角）
        current = node
        v = leaf_value
        while current is not None:
            current.visits += 1
            current.value_sum += v
            v = -v  # 翻转视角：我的好事是对手的坏事
            current = current.parent

    # 返回每个走法的访问次数分布
    visit_counts = {}
    for move in board.legal_moves():
        if move in root.children:
            visit_counts[move] = root.children[move].visits
        else:
            visit_counts[move] = 0

    return visit_counts


# ============================================================
# 第四部分：自我对弈 + 训练
# ============================================================

def self_play(net, num_simulations=50):
    """
    让 AI 跟自己对弈一局，收集训练数据

    返回：[(state, mcts_policy, winner), ...]
      - state: 棋盘状态
      - mcts_policy: MCTS 搜索后的走法概率分布（比神经网络原始输出更准）
      - winner: 最终赢家
    """
    board = Board()
    game_data = []

    while True:
        # 用 MCTS 搜索当前最佳走法
        visit_counts = mcts_search(board, net, num_simulations)

        # 把访问次数转成概率分布
        total_visits = sum(visit_counts.values())
        if total_visits == 0:
            break
        policy = np.zeros(9)
        for move, count in visit_counts.items():
            policy[move] = count / total_visits

        # 记录训练数据
        game_data.append((board.to_tensor().copy(), policy.copy(), board.current_player))

        # 选访问次数最多的走法（贪心）
        best_move = max(visit_counts, key=visit_counts.get)
        board.play(best_move)

        # 检查游戏是否结束
        winner = board.check_winner()
        if winner is not None:
            break

    # 把 winner 转成每个训练样本的标签
    training_data = []
    for state, policy, player in game_data:
        if winner == 0:
            value = 0  # 平局
        elif winner == player:
            value = 1  # 我赢了
        else:
            value = -1  # 我输了
        training_data.append((state, policy, value))

    return training_data, winner


def train(num_games=200, num_simulations=50, lr=0.003):
    """
    完整的训练流程：
    1. 自我对弈生成数据
    2. 用数据训练神经网络
    3. 重复，越来越强
    """
    net = PolicyValueNet()
    optimizer = optim.Adam(net.parameters(), lr=lr)

    print("=" * 50)
    print("AlphaGo 井字棋 - 开始训练")
    print("=" * 50)

    for game_idx in range(num_games):
        # --- 自我对弈 ---
        data, winner = self_play(net, num_simulations)

        # --- 用对弈数据训练网络 ---
        total_loss = 0
        for state, policy, value in data:
            state_t = torch.FloatTensor(state).unsqueeze(0)
            policy_t = torch.FloatTensor(policy).unsqueeze(0)
            value_t = torch.FloatTensor([value]).unsqueeze(0)

            # 前向传播
            pred_policy, pred_value = net(state_t)

            # 损失函数 = 策略损失 + 价值损失
            # 策略损失：让网络输出接近 MCTS 的搜索分布
            policy_loss = -torch.sum(policy_t * torch.log_softmax(pred_policy, dim=1))
            # 价值损失：让网络评估接近真实胜负
            value_loss = (pred_value - value_t) ** 2

            loss = policy_loss + value_loss

            # 反向传播，更新权重
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            total_loss += loss.item()

        # 每 20 局打印一次进度
        if (game_idx + 1) % 20 == 0:
            result_str = {1: "黑赢", -1: "白赢", 0: "平局"}[winner]
            print(f"第 {game_idx+1:3d} 局 | 结果: {result_str} | 平均损失: {total_loss/len(data):.4f}")

    print("=" * 50)
    print("训练完成！")
    return net


# ============================================================
# 第五部分：人机对弈
# ============================================================

def play_against_human(net, num_simulations=200):
    """和训练好的 AI 下一盘"""
    board = Board()
    print("你是 X（先手），AI 是 O（后手）")
    print("输入 0-8 落子：")
    print("0 1 2")
    print("3 4 5")
    print("6 7 8")
    print()
    board.display()

    while True:
        # 人类走
        if board.current_player == 1:
            while True:
                try:
                    move = int(input(f"你的走法 (合法: {board.legal_moves()}): "))
                    if move in board.legal_moves():
                        break
                    print("非法走法，重试")
                except ValueError:
                    print("输入数字 0-8")
            board.play(move)
        else:
            # AI 走
            print("AI 思考中...")
            visit_counts = mcts_search(board, net, num_simulations)
            best_move = max(visit_counts, key=visit_counts.get)
            print(f"AI 选择位置 {best_move}")
            board.play(best_move)

        board.display()

        winner = board.check_winner()
        if winner is not None:
            if winner == 1:
                print("你赢了！")
            elif winner == -1:
                print("AI 赢了！")
            else:
                print("平局！")
            break


# ============================================================
# 主程序
# ============================================================

if __name__ == "__main__":
    # 训练
    net = train(num_games=200, num_simulations=50)

    # 保存模型
    torch.save(net.state_dict(), "tictactoe_alphago.pth")
    print("\n模型已保存到 tictactoe_alphago.pth")

    # 人机对弈
    print("\n" + "=" * 50)
    print("来和 AI 下一盘吧！")
    print("=" * 50)
    play_against_human(net)

File: alphago_tictactoe.js
/**
 * AlphaGo 简化版 —— 井字棋 (Node.js + TensorFlow.js)
 * ====================================================
 * 核心组件：
 * 1. 棋盘逻辑（Board）
 * 2. 神经网络（PolicyValueNet）—— 策略 + 价值
 * 3. 蒙特卡洛树搜索（MCTS）—— 用神经网络指导搜索
 * 4. 自我对弈训练（train）—— 通过实战提升棋力
 *
 * 运行: npm install && node alphago_tictactoe.js
 */

const tf = require('@tensorflow/tfjs');
const fs = require('fs');
const readline = require('readline');

// ============================================================
// 第一部分：棋盘逻辑
// ============================================================

class Board {
  constructor() {
    // 3x3 棋盘，0=空，1=黑棋(先手)，-1=白棋(后手)
    this.grid = new Float32Array(9);
    this.currentPlayer = 1; // 黑棋先走
  }

  copy() {
    const b = new Board();
    b.grid = new Float32Array(this.grid);
    b.currentPlayer = this.currentPlayer;
    return b;
  }

  legalMoves() {
    const moves = [];
    for (let i = 0; i < 9; i++) {
      if (this.grid[i] === 0) moves.push(i);
    }
    return moves;
  }

  play(pos) {
    if (this.grid[pos] !== 0) throw new Error(`位置 ${pos} 已有棋子`);
    this.grid[pos] = this.currentPlayer;
    this.currentPlayer *= -1; // 换手
  }

  checkWinner() {
    const lines = [
      [0,1,2], [3,4,5], [6,7,8], // 横
      [0,3,6], [1,4,7], [2,5,8], // 竖
      [0,4,8], [2,4,6],          // 对角
    ];
    for (const [a, b, c] of lines) {
      const s = this.grid[a] + this.grid[b] + this.grid[c];
      if (s === 3) return 1;    // 黑赢
      if (s === -3) return -1;  // 白赢
    }
    if (this.legalMoves().length === 0) return 0; // 平局
    return null; // 未结束
  }

  toTensor() {
    // 3个通道：我的棋子、对手棋子、当前玩家标识
    const myPieces = new Float32Array(9);
    const oppPieces = new Float32Array(9);
    const playerPlane = new Float32Array(9);

    for (let i = 0; i < 9; i++) {
      myPieces[i] = this.grid[i] === this.currentPlayer ? 1 : 0;
      oppPieces[i] = this.grid[i] === -this.currentPlayer ? 1 : 0;
      playerPlane[i] = this.currentPlayer;
    }

    // 拼接成 27 维向量
    const result = new Float32Array(27);
    result.set(myPieces, 0);
    result.set(oppPieces, 9);
    result.set(playerPlane, 18);
    return result;
  }

  display() {
    const symbols = { 0: '.', 1: 'X', '-1': 'O' };
    for (let row = 0; row < 3; row++) {
      const line = [];
      for (let col = 0; col < 3; col++) {
        line.push(symbols[this.grid[row * 3 + col]] || '.');
      }
      console.log(line.join(' '));
    }
    console.log();
  }
}

// ============================================================
// 第二部分：神经网络 —— 策略头 + 价值头
// ============================================================

function createPolicyValueNet() {
  // 共享特征提取层
  const input = tf.input({ shape: [27] });
  let x = tf.layers.dense({ units: 128, activation: 'relu' }).apply(input);
  x = tf.layers.dense({ units: 128, activation: 'relu' }).apply(x);

  // 策略头：输出每个位置的落子概率（9个 logits）
  const policyLogits = tf.layers.dense({ units: 9, name: 'policy' }).apply(x);

  // 价值头：输出当前局面胜率（tanh 压缩到 [-1, 1]）
  const valueRaw = tf.layers.dense({ units: 1, name: 'value' }).apply(x);
  const value = tf.layers.activation({ activation: 'tanh' }).apply(valueRaw);

  const model = tf.model({ inputs: input, outputs: [policyLogits, value] });
  return model;
}

/**
 * 用神经网络评估局面
 * 返回 { policy: Float32Array(9), value: number }
 */
function evaluatePosition(model, stateArray) {
  return tf.tidy(() => {
    const input = tf.tensor2d([Array.from(stateArray)]);
    const [policyLogits, valueTensor] = model.predict(input);
    const policy = tf.softmax(policyLogits, 1);
    return {
      policy: policy.dataSync(),
      value: valueTensor.dataSync()[0]
    };
  });
}

// ============================================================
// 第三部分：蒙特卡洛树搜索（MCTS）
// ============================================================

class MCTSNode {
  constructor(parent = null, move = null) {
    this.parent = parent;
    this.move = move;
    this.children = new Map(); // move -> MCTSNode
    this.visits = 0;
    this.valueSum = 0;
    this.prior = 0;
  }

  get value() {
    return this.visits > 0 ? this.valueSum / this.visits : 0;
  }

  ucbScore(cPuct = 1.4) {
    if (this.visits === 0) {
      return Infinity;
    }
    const parentVisits = this.parent ? this.parent.visits : 1;
    const exploration = cPuct * this.prior * Math.sqrt(parentVisits) / (1 + this.visits);
    return this.value + exploration;
  }
}

function mctsSearch(board, model, numSimulations = 100) {
  const root = new MCTSNode();

  for (let sim = 0; sim < numSimulations; sim++) {
    let node = root;
    const simBoard = board.copy();

    // --- 1. 选择：沿 UCB 最高的路径走到叶子 ---
    while (node.children.size > 0 && simBoard.checkWinner() === null) {
      let bestMove = null;
      let bestUcb = -Infinity;
      for (const [move, child] of node.children) {
        const ucb = child.ucbScore();
        if (ucb > bestUcb) {
          bestUcb = ucb;
          bestMove = move;
        }
      }
      node = node.children.get(bestMove);
      simBoard.play(bestMove);
    }

    // --- 2. 扩展：评估叶子节点 ---
    const winner = simBoard.checkWinner();
    let leafValue;

    if (winner !== null) {
      // 游戏已结束，用真实结果
      if (winner === 0) {
        leafValue = 0;
      } else if (winner === board.currentPlayer) {
        leafValue = 1;
      } else {
        leafValue = -1;
      }
    } else {
      // 用神经网络评估
      const state = simBoard.toTensor();
      const { policy, value } = evaluatePosition(model, state);
      leafValue = value;

      // 创建子节点，用神经网络的策略作为先验概率
      const legal = simBoard.legalMoves();

      // 归一化合法走法的概率
      const legalPolicy = {};
      let total = 0;
      for (const m of legal) {
        legalPolicy[m] = policy[m];
        total += policy[m];
      }
      if (total > 0) {
        for (const m of legal) {
          legalPolicy[m] /= total;
        }
      } else {
        for (const m of legal) {
          legalPolicy[m] = 1 / legal.length;
        }
      }

      for (const move of legal) {
        const child = new MCTSNode(node, move);
        child.prior = legalPolicy[move];
        node.children.set(move, child);
      }
    }

    // --- 3. 回溯：把叶子价值传回根节点 ---
    let current = node;
    let v = leafValue;
    while (current !== null) {
      current.visits += 1;
      current.valueSum += v;
      v = -v; // 翻转视角
      current = current.parent;
    }
  }

  // 返回每个走法的访问次数分布
  const visitCounts = {};
  for (const move of board.legalMoves()) {
    if (root.children.has(move)) {
      visitCounts[move] = root.children.get(move).visits;
    } else {
      visitCounts[move] = 0;
    }
  }

  return visitCounts;
}

// ============================================================
// 第四部分：自我对弈 + 训练
// ============================================================

function selfPlay(model, numSimulations = 50, verbose = false, logFile = null, gameIdx = null) {
  function log(msg) {
    if (verbose) console.log(msg);
    if (logFile) {
      fs.appendFileSync(logFile, msg + '\n');
    }
  }

  const board = new Board();
  const gameData = [];
  const moveHistory = [];

  if (verbose || logFile) {
    log(`\n${'='.repeat(50)}`);
    log(`第 ${gameIdx} 局开始`);
    log(`${'='.repeat(50)}`);
  }

  let step = 0;
  let winner;

  while (true) {
    step++;
    const playerName = board.currentPlayer === 1 ? 'X(黑)' : 'O(白)';

    // 用 MCTS 搜索当前最佳走法
    const visitCounts = mctsSearch(board, model, numSimulations);

    // 把访问次数转成概率分布
    const totalVisits = Object.values(visitCounts).reduce((a, b) => a + b, 0);
    if (totalVisits === 0) break;

    const policy = new Float32Array(9);
    for (const [move, count] of Object.entries(visitCounts)) {
      policy[parseInt(move)] = count / totalVisits;
    }

    // 记录训练数据
    gameData.push({
      state: board.toTensor().slice(),
      policy: policy.slice(),
      player: board.currentPlayer
    });

    // 选访问次数最多的走法
    let bestMove = null;
    let maxVisits = -1;
    for (const [move, count] of Object.entries(visitCounts)) {
      if (count > maxVisits) {
        maxVisits = count;
        bestMove = parseInt(move);
      }
    }

    // 日志
    if (verbose || logFile) {
      log(`\n--- 第 ${step} 步 | ${playerName} 行棋 ---`);
      log('当前棋盘:');
      const gridDisplay = [];
      for (let i = 0; i < 9; i++) {
        if (board.grid[i] === 1) gridDisplay.push('X');
        else if (board.grid[i] === -1) gridDisplay.push('O');
        else gridDisplay.push(String(i));
      }
      log(`  ${gridDisplay[0]} | ${gridDisplay[1]} | ${gridDisplay[2]}`);
      log(`  ---------`);
      log(`  ${gridDisplay[3]} | ${gridDisplay[4]} | ${gridDisplay[5]}`);
      log(`  ---------`);
      log(`  ${gridDisplay[6]} | ${gridDisplay[7]} | ${gridDisplay[8]}`);

      log(`MCTS 搜索分布 (共 ${totalVisits} 次模拟):`);
      const sortedMoves = Object.entries(visitCounts).sort((a, b) => b[1] - a[1]);
      for (const [move, count] of sortedMoves) {
        const pct = (count / totalVisits * 100).toFixed(1);
        const bar = '█'.repeat(Math.floor(pct / 2));
        const marker = parseInt(move) === bestMove ? ' ← 选中' : '';
        log(`  位置 ${move}: ${String(count).padStart(4)} 次 (${pct.padStart(5)}%) ${bar}${marker}`);
      }

      // 神经网络先验
      const { policy: netPolicy, value: netValue } = evaluatePosition(model, board.toTensor());
      log(`神经网络评估: 胜率预测 ${netValue >= 0 ? '+' : ''}${netValue.toFixed(2)}`);
      log(`神经网络先验 (Top 3):`);
      const top3 = Array.from({ length: 9 }, (_, i) => i).sort((a, b) => netPolicy[b] - netPolicy[a]).slice(0, 3);
      for (const i of top3) {
        log(`  位置 ${i}: ${(netPolicy[i] * 100).toFixed(1)}%`);
      }
    }

    board.play(bestMove);
    moveHistory.push(`${playerName}@${bestMove}`);

    if (verbose || logFile) {
      log(`决策: ${playerName} 落子位置 ${bestMove}`);
    }

    winner = board.checkWinner();
    if (winner !== null) break;
  }

  // 记录对局结果
  if (verbose || logFile) {
    log(`\n${'─'.repeat(50)}`);
    const result = winner === 1 ? 'X(黑) 胜' : winner === -1 ? 'O(白) 胜' : '平局';
    log(`对局结果: ${result} (共 ${step} 步)`);
    log(`走法记录: ${moveHistory.join(' → ')}`);
    log(`${'─'.repeat(50)}`);
  }

  // 把 winner 转成每个训练样本的标签
  const trainingData = [];
  for (const { state, policy, player } of gameData) {
    let value;
    if (winner === 0) value = 0;
    else if (winner === player) value = 1;
    else value = -1;
    trainingData.push({ state, policy, value });
  }

  return { trainingData, winner };
}

// ============================================================
// 训练函数
// ============================================================

async function train(numGames = 200, numSimulations = 50, lr = 0.003, verbose = false, logPath = 'training_log.txt') {
  const model = createPolicyValueNet();

  // 使用 Adam 优化器
  const optimizer = tf.train.adam(lr);

  // 初始化日志文件
  fs.writeFileSync(logPath,
    `AlphaGo 井字棋训练日志\n` +
    `训练参数: ${numGames} 局, ${numSimulations} 次模拟/步, lr=${lr}\n` +
    `开始时间: ${new Date().toLocaleString()}\n` +
    `${'='.repeat(50)}\n`
  );

  console.log('='.repeat(50));
  console.log('AlphaGo 井字棋 - 开始训练');
  console.log(`日志文件: ${logPath}`);
  console.log('='.repeat(50));

  const stats = { blackWin: 0, whiteWin: 0, draw: 0 };
  const startTime = Date.now();

  for (let gameIdx = 0; gameIdx < numGames; gameIdx++) {
    const gameStart = Date.now();

    // 自我对弈
    const showVerbose = verbose || gameIdx < 5 || (gameIdx + 1) % 10 === 0;
    const { trainingData, winner } = selfPlay(
      model, numSimulations,
      showVerbose && verbose,
      logPath,
      gameIdx + 1
    );

    // 用对弈数据训练网络
    let totalLoss = 0;

    for (const { state, policy, value } of trainingData) {
      const loss = optimizer.minimize(() => {
        return tf.tidy(() => {
          const stateT = tf.tensor2d([Array.from(state)]);
          const policyT = tf.tensor2d([Array.from(policy)]);
          const valueT = tf.tensor2d([[value]]);

          const [predPolicy, predValue] = model.apply(stateT);

          // 策略损失：交叉熵
          const logProbs = tf.logSoftmax(predPolicy, 1);
          const policyLoss = tf.neg(tf.sum(tf.mul(policyT, logProbs), 1));

          // 价值损失：MSE
          const valueLoss = tf.square(tf.sub(predValue, valueT));

          const totalLoss = tf.add(policyLoss, valueLoss);
          return totalLoss.mean();
        });
      }, true); // returnCost = true

      totalLoss += loss.dataSync()[0];
      loss.dispose();
    }

    // 更新统计
    const gameTime = (Date.now() - gameStart) / 1000;
    if (winner === 1) stats.blackWin++;
    else if (winner === -1) stats.whiteWin++;
    else stats.draw++;

    // 每 10 局打印进度
    if ((gameIdx + 1) % 10 === 0) {
      const resultStr = { 1: '黑赢', '-1': '白赢', 0: '平局' }[winner];
      const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
      console.log(
        `第 ${String(gameIdx + 1).padStart(3)}/${numGames} 局 | ` +
        `结果: ${resultStr} | ` +
        `损失: ${(totalLoss / trainingData.length).toFixed(4)} | ` +
        `耗时: ${gameTime.toFixed(1)}s | ` +
        `统计: 黑${stats.blackWin} 白${stats.whiteWin} 平${stats.draw}`
      );

      fs.appendFileSync(logPath,
        `\n[摘要] 第 ${gameIdx + 1} 局 | ` +
        `损失: ${(totalLoss / trainingData.length).toFixed(4)} | ` +
        `耗时: ${gameTime.toFixed(1)}s | ` +
        `累计: 黑${stats.blackWin} 白${stats.whiteWin} 平${stats.draw}\n`
      );
    }
  }

  const totalTime = ((Date.now() - startTime) / 1000).toFixed(1);
  const summary =
    `\n${'='.repeat(50)}\n` +
    `训练完成！\n` +
    `总耗时: ${totalTime}s\n` +
    `战绩统计: 黑赢 ${stats.blackWin} | 白赢 ${stats.whiteWin} | 平局 ${stats.draw}\n` +
    `${'='.repeat(50)}`;

  console.log(summary);
  fs.appendFileSync(logPath, summary);
  console.log(`详细日志已保存到: ${logPath}`);

  return model;
}

// ============================================================
// 第五部分：人机对弈
// ============================================================

async function playAgainstHuman(model, numSimulations = 200) {
  const board = new Board();
  console.log('你是 X（先手），AI 是 O（后手）');
  console.log('输入 0-8 落子：');
  console.log('0 1 2');
  console.log('3 4 5');
  console.log('6 7 8');
  console.log();
  board.display();

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const askQuestion = (query) => new Promise(resolve => rl.question(query, resolve));

  while (true) {
    if (board.currentPlayer === 1) {
      // 人类走
      let move;
      while (true) {
        const answer = await askQuestion(`你的走法 (合法: [${board.legalMoves().join(', ')}]): `);
        move = parseInt(answer);
        if (!isNaN(move) && board.legalMoves().includes(move)) break;
        console.log('非法走法，重试');
      }
      board.play(move);
    } else {
      // AI 走
      console.log('AI 思考中...');
      const visitCounts = mctsSearch(board, model, numSimulations);
      let bestMove = null;
      let maxVisits = -1;
      for (const [move, count] of Object.entries(visitCounts)) {
        if (count > maxVisits) {
          maxVisits = count;
          bestMove = parseInt(move);
        }
      }
      console.log(`AI 选择位置 ${bestMove}`);
      board.play(bestMove);
    }

    board.display();

    const winner = board.checkWinner();
    if (winner !== null) {
      if (winner === 1) console.log('你赢了！');
      else if (winner === -1) console.log('AI 赢了！');
      else console.log('平局！');
      break;
    }
  }

  rl.close();
}

// ============================================================
// 模型保存/加载
// ============================================================

async function saveModel(model, path = './tictactoe_model') {
  await model.save(`file://${path}`);
  console.log(`模型已保存到 ${path}/`);
}

async function loadModel(path = './tictactoe_model') {
  const model = await tf.loadLayersModel(`file://${path}/model.json`);
  console.log(`已加载模型 ${path}/`);
  return model;
}

// ============================================================
// 主程序
// ============================================================

async function main() {
  const args = process.argv.slice(2);
  const verbose = args.includes('--verbose');
  const playOnly = args.includes('--play');

  if (playOnly) {
    const model = await loadModel();
    await playAgainstHuman(model);
  } else {
    // 训练
    const model = await train(
      200,   // numGames
      50,    // numSimulations
      0.003, // lr
      verbose,
      'training_log.txt'
    );

    // 保存模型
    await saveModel(model);

    // 人机对弈
    console.log('\n' + '='.repeat(50));
    console.log('来和 AI 下一盘吧！');
    console.log('='.repeat(50));
    await playAgainstHuman(model);
  }
}

// 导出模块（供测试用）
if (typeof module !== 'undefined') {
  module.exports = { Board, MCTSNode, createPolicyValueNet, mctsSearch, selfPlay, train, playAgainstHuman };
}

if (require.main === module) {
  main().catch(console.error);
}
